## Lab 2: Clustering ##

The dataset for this lab has been created from some custom features from Lab 1. The columns are named as q1, q2....etc. A custom description of the features can be found at this link: (https://drive.google.com/file/d/0BycRP8EC1WVVUmo5d3kwTWlDYlk/view). 

If, at any place, you wish to include a comment or a discussion of your code/results, please add a Text cell below your answer/code. A text cell can be added by clicking on '+TEXT' in the upper left corner.


#Environment Setup
Run this cell to setup your environment.

In [57]:
import pandas as pd
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
import numpy as np
from sklearn.preprocessing import normalize

import matplotlib
import matplotlib.pyplot as plt

%matplotlib inline
matplotlib.style.use('ggplot')



#DOWNLOADING DATASET
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/yelp_reviewers.zip
!unzip yelp_reviewers.zip
print('Dataset Downloaded: yelp_reviewers.csv')
df = pd.read_csv('yelp_reviewers.csv')
print(df.shape)
df.dropna().describe()

print('....SETUP COMPLETE....')

--2019-02-14 19:33:07--  http://people.ischool.berkeley.edu/~zp/course_datasets/yelp_reviewers.zip
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12350863 (12M) [application/zip]
Saving to: ‘yelp_reviewers.zip.2’

yelp_reviewers.zip. 100%[===================>]  11.78M  2.21MB/s    in 6.3s    

2019-02-14 19:33:13 (1.86 MB/s) - ‘yelp_reviewers.zip.2’ saved [12350863/12350863]

Archive:  yelp_reviewers.zip
replace yelp_reviewers.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: yelp_reviewers.csv      
Dataset Downloaded: yelp_reviewers.csv
(171639, 43)
....SETUP COMPLETE....


### Question 1 ###

Choose an implementation of k-means. Which one have you used? (Scikit-learn, custom, other)



### ANSWER: Scikit-learn ### 

In [22]:
df.head()

user_id  q3  q4  q5  q6    q7  q8  q9  q10  q11    ...     \
0  --1Y03CEKR3WDbBjYnsW7A   1   0   0   0  0.00 NaN NaN  NaN  NaN    ...      
1  --2QZsyXGz1OhiD4-0FQLQ   3   0   0   1  1.10 NaN NaN  0.0  0.0    ...      
2  --82_AVgRBsLw6Dhy8sEnA   1   0   0   0  0.00 NaN NaN  NaN  NaN    ...      
3  --8A9o_NeGyt_3kzlXtSdg   3   0   0   1  1.10 NaN NaN  0.0  0.0    ...      
4  --8WbseBk1NjfPiZWjQ-XQ   4   0   0   0  1.39 NaN NaN  NaN  NaN    ...      

   q16t  q16u  q16v  q16w      q16x        q16y  q16z  q16aa     q16ab  \
0    no  4.00    14     5  0.000000  213.000000     0      0  0.000000   
1    no  1.00    19    14  0.000000   21.833333     1      0  0.333333   
2    no  3.00    31     4  0.002275  165.000000     0      0  0.000000   
3    no  0.00     8    11  0.002242   61.500000     0      0  0.333333   
4    no  0.25    28    11  0.000444  287.500000     0      0  0.000000   

      q16ac  
0  5.000000  
1  4.666667  
2  4.000000  
3  3.666667  
4  2.750000  

[5 rows x 43 columns]

In [26]:
df_s = df.sample(n = int(171639*0.3))
print(df_s.shape)

(51491, 43)




---


### Question 2 ###
What is the best choice of k according to the silhouette metric for clustering q4-q6 (# of cool, funny, useful votes). Only consider 2 <= k <= 8.  

NOTE: For features with high variance, empty clusters can occur. There are several ways of dealing with empty clusters. A common approach is to drop empty clusters, the prefered approach for this Lab is to treat the empty cluster as a “singleton” leaving it empty with a single point placeholder.


In [27]:
# YOUR CODE HERE

Y = df_s[['q4','q5','q6']]

for k in range(2,9):
    kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)
    s_score = silhouette_score(Y, kmeans.labels_)
    print(k, s_score)

2 0.9863428851460094
3 0.9746234578068156
4 0.9108376688107326
5 0.891430312517317
6 0.7909151937480288
7 0.7450647471631096
8 0.6938218320393711


### ANSWER: k=2 ###

***

### Question 3 ###
Answer question 2 but using the log of the features (q7-q10)



In [28]:
# Drop NA

df_a = df_s[['q7','q8','q9','q10']]

Y = df_a.dropna()

for k in range(2,9):
    kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)
    s_score = silhouette_score(Y, kmeans.labels_)
    print(k, s_score)

2 0.4213258252228127
3 0.3228265206739778
4 0.271116735487955
5 0.2732222159140612
6 0.24682273359576626
7 0.232682481836789
8 0.2334093669124342


In [29]:
# Fill NA with 0

df_s_f = df_s
df_s_f.fillna(value=0, inplace=True)

Y = df_s_f[['q7','q8','q9','q10']]

for k in range(2,9):
    kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)
    s_score = silhouette_score(Y, kmeans.labels_)
    print(k, s_score)

2 0.6113801079208747
3 0.4919416840278804
4 0.5344306059072813
5 0.5421128561041991
6 0.5660360991006554
7 0.5763846849527425
8 0.58099167690553


### ANSWER: k = 2 ###

***

### Question 4 ###
Answer question 2 but using the percentage of the features (q11-q13)



In [30]:
# Drop NA

df_b = df_s[['q11','q12','q13']]

Y = df_b.dropna()

for k in range(2,9):
    kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)
    s_score = silhouette_score(Y, kmeans.labels_)
    print(k, s_score)

2 0.5691653002288268
3 0.6760273932075892
4 0.7234589653737166
5 0.7615932260948528
6 0.758455640002874
7 0.7725891759027274
8 0.7897932310693094


In [31]:
# Fill NA with 0 since these are all ratios, NA may be the denominator (total votes) = 0

Y = df_s_f[['q11','q12','q13']]

for k in range(2,9):
    kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)
    s_score = silhouette_score(Y, kmeans.labels_)
    print(k, s_score)

2 0.5691653002288268
3 0.6760273932075892
4 0.7234589653737166
5 0.7615932260948528
6 0.758455640002874
7 0.7725891759027274
8 0.7897932310693094


### ANSWER: k = 8 ###

***

### Question 5 ###
Inspect the [best] clustering generated from question 4

#### 5.a ####
**Question**: List the number of data points in each cluster (eg. C1: 2,000 C2: 4,200 etc)

In [35]:
# YOUR CODE HERE

df_f = df
df_f.fillna(value=0, inplace=True)

Y = df_f[['q11','q12','q13']]
k = 8

kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)

list_label = kmeans.labels_.tolist()

c1 = list_label.count(0)
c2 = list_label.count(1)
c3 = list_label.count(2)
c4 = list_label.count(3)
c5 = list_label.count(4)
c6 = list_label.count(5)
c7 = list_label.count(6)
c8 = list_label.count(7)

print( "C1 : {}, C2 : {}, C3 : {}, C4 : {}, C5 : {}, C6 : {}, C7 : {}, C8 : {}. ".format(c1,c2,c3,c4,c5,c6,c7,c8) )

C1 : 18958, C2 : 33153, C3 : 71282, C4 : 10042, C5 : 18074, C6 : 5811, C7 : 4026, C8 : 10293. 


#### 5.b ####
**Question**: Were there clusters that represented very funny but useless reviewers?  If so, print them.

**Answer**:  

In [42]:
# Means of each group (1~8)

df['cluster'] = kmeans.labels_
mean = df.groupby('cluster').mean()
df_mean = pd.DataFrame()
df_mean['q11'] = mean['q11']
df_mean['q12'] = mean['q12']
df_mean['q13'] = mean['q13']

print(df_mean)


               q11        q12        q13
cluster                                 
0        31.596688  30.939013  37.461630
1         0.283142   0.393800  99.323041
2         0.000000   0.000000   0.000000
3         3.703326  44.456666  51.840012
4        21.056556  10.766804  68.177308
5         1.386662  97.049640   1.563829
6        98.205656   0.899779   0.894615
7        48.994693   2.325515  48.679865


In [45]:
# Pick the one with high q12 but low q13
df_fun_useless = df[df['cluster']==5]
print(df_fun_useless['user_id'])

10        --D6B8SF9EAIaS4O7wLVJw
13        --JC6ri9lh1-tazGOFA3yg
29        --mUxQG05DwbTd_Y8uCAFw
39        -0-VtlGnGqOcTwDK-WivBQ
175       -3Ap7ootO4TIu4k725_6sQ
199       -3eJTCKXSDY6ZvIng-jIoA
213       -3vG-bLDopHV-q5v5mmRxg
299       -6KDFG8HN1xldJE1pGTnwg
317       -6kx-CN42cAvXgO8qiS8EA
318       -6lE7d3bk-I8x5qQTfkGLQ
320       -6o9ZP_9SQznWR9Jz-bP2g
365       -7iKlPj7TIBKaY-5OANi9g
372       -7u0ORuRj5nRPxM3Do9axA
387       -8ES3xZyJuBUMmFNyzxKgA
485       -AYwTUHKlKK5B_gtGba3UQ
501       -B32HlBjDRafJpD_eW5pxw
510       -BFtZw3FybgNEwzoWljQ3Q
520       -B_wKTfcRxGjchsqsc4fPA
522       -Bbs0C1lzVvW6-5OYp_yUw
548       -CZ3BYqA7xbcti0gWODWkw
611       -ECKlfeVcmBSY-Wdt-oPDQ
625       -ESkYkhwSsB1DeflMVJhXA
639       -EkwH-zZsiYmSxbFEYWT4g
656       -FFFkkCjDVsujL-4UoLASA
658       -FIf6h5k9wYMea5v7gukqg
659       -FJPYc82-vJZGUBrKCUPuQ
661       -FP5UB8n-LtBjJ1Zwj95ow
750       -HYD2dWhxU6oAr4w3BPLtQ
793       -ImyzB4PNtxQNx8Rs7Cedg
799       -IpbvH7EWk6yUjrxDfYaKQ
          

### ANSWER: Yes. It's c6 (cluster==5) ###
### c6 is the one with lowest useful vote rate (97.0%), and highest funny vote rate (1.5%) ###

#### 5.c ####
**Question**: How many reviewers were in the cluster that represented relatively equal strength in all voting categories (assuming such a cluster exists in your clustering)?  

**Answer**:

In [43]:
# Relatively equal votes

print(len(df[df['cluster']==0]))
print(len(df[df['cluster']==2]))

18958
71282


### ANSWER: The groups that with relatively equal votes are C1 and C3, and their number of reviewers are 18958 and 71282 respectively. ###

***

### Question 6 ###
Cluster the dataset using $k = 5$ and using features q7-q13 (log and % type votes) and q14 (most active year feature) and the natural log of q15 (avg review chars)

In [0]:
# YOUR CODE HERE

import math

df_c = df_f.loc[ (df_f['q15']!=0) ]
df_c['log15'] = df_c['q15'].apply(lambda x: math.log(x))
Y = df_c[['q7','q8','q9','q10','q11','q12','q13','q14','log15']]

k = 5
kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)

### ANSWER: ###

#### 6.a ####
**Question**: What is the silhouette metric for this clustering?   
You may use the max, as you did in question 2. For a more in-depth understanding of cluster analysis with silhouette, look [here](http://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html)

In [48]:
# YOUR CODE HERE
s_score = silhouette_score(Y, kmeans.labels_)
print(s_score)

0.7350661974287377


### ANSWER: ###

#### 6.b ####
**Question**: What was the average “number of reviews per reviewer (q3)” among the points in each of the clusters (eg. C1: 1.4 C2: 4.2 C3: 3.4 etc)

In [54]:
# YOUR CODE HERE

df_c['cluster_2'] = kmeans.labels_
mean = df_c.groupby('cluster_2').mean()
print(mean[['q3']])

                 q3
cluster_2          
0          2.810288
1          5.457982
2          1.683875
3          3.224257
4          2.645331


### ANSWER: ###

***

### Question 7 ###
Cluster the dataset using the features described in question 6 + every group’s question 16 features (you may drop features with high incidents of -Inf / blank / or NaN values). It is suggested that you perform some form of normalization on these question 16 features so as not to over bias the clustering towards the larger magnitude features.

#### Data Cleansing and Normalization ####
Check how many null values there are in each column.

In [58]:
# YOUR CODE HERE
df.isna().sum()

user_id         0
q3              0
q4              0
q5              0
q6              0
q7              0
q8         117853
q9         122162
q10         81153
q11         71282
q12         71282
q13         71282
q14             0
q15             0
q16a            0
q16b            0
q16c            0
q16d            0
q16e            0
q16f            0
q16g            0
q16h            0
q16i            0
q16j            0
q16k            0
q16l            0
q16m            0
q16n            0
q16o            0
q16p            0
q16q            0
q16r            0
q16s            0
q16t            0
q16u            0
q16v            0
q16w            0
q16x            0
q16y            0
q16z            0
q16aa           0
q16ab       47594
q16ac           0
dtype: int64

It looks like q8 - q13 and q16ab have a lot of null values, especially q8 and q9. Let's see what the impact is of removing some of these columnsbefore removing any columns

In [78]:
# YOUR CODE HERE

df_1 = df.drop(columns = ['q3','q4','q5','q6'])
df_1 = df_1.dropna()

df_2 = df.drop(columns = ['q3','q4','q5','q6',   'q8','q9'])
df_2 = df_2.dropna()

print( len( df_1 ) )
print( len( df_2 ) )

24205
65546


By removing 2 features, we double the number of rows remaining. That's pretty good.  
Preprocess categorical variables to dummy values.

In [79]:
# Check column type

df_2.dtypes

user_id     object
q7         float64
q10        float64
q11        float64
q12        float64
q13        float64
q14          int64
q15        float64
q16a         int64
q16b       float64
q16c       float64
q16d       float64
q16e       float64
q16f       float64
q16g         int64
q16h         int64
q16i         int64
q16j       float64
q16k         int64
q16l         int64
q16m       float64
q16n       float64
q16o       float64
q16p       float64
q16q       float64
q16r         int64
q16s        object
q16t        object
q16u       float64
q16v         int64
q16w         int64
q16x       float64
q16y       float64
q16z         int64
q16aa        int64
q16ab      float64
q16ac      float64
dtype: object

In [0]:
df_3 = pd.get_dummies(df_2[['q16s','q16t']])

Now normalize the remaining features.

In [83]:
df_4 = df_2.drop(columns = ['user_id','q16s','q16t'])
df_4['log15'] = df_4['q15'].apply(lambda x: math.log(x))
df_4 = df_4.drop(columns = ['q15'])

normalized_df=(df_4 - df_4.mean()) / df_4.std()

df_5 = pd.concat([normalized_df, df_3], axis=1)

print(df_5.head())

         q7       q10       q11       q12       q13       q14      q16a  \
1 -0.008373 -0.970540 -0.923858 -0.794090  1.190034  0.338026 -0.671293   
3 -0.008373 -0.970540 -0.923858 -0.794090  1.190034  0.790749 -0.005596   
5 -0.008373  0.200271 -0.923858 -0.794090  1.190034  0.338026 -0.671293   
6  1.144595  0.934690  0.457137 -0.794090  0.221562  0.338026  2.657190   
9 -0.490737 -0.970540 -0.923858  2.043068 -0.746910  0.338026 -0.005596   

       q16b      q16c      q16d    ...         q16y      q16z     q16aa  \
1 -0.346419 -0.134327  0.946036    ...    -0.923302  0.571247 -0.238609   
3  1.768061 -0.235786  0.125263    ...    -0.193118 -0.385224 -0.238609   
5  1.768061 -0.105841  0.125263    ...    -0.570482 -0.385224 -0.238609   
6  1.384890 -0.116397 -0.114130    ...    -0.549198 -0.385224 -0.238609   
9  1.538453 -0.096514 -0.011533    ...    -0.280556 -0.385224 -0.238609   

      q16ab     q16ac     log15  q16s_experienced  q16s_freshman  q16t_no  \
1 -0.279577  0.946036

#### 7.a ####
**Question**: Using the silhouette metric, what was the best k?  

**Answer**:  

In [84]:
Y = df_5

for k in range(2,9):
    kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)
    s_score = silhouette_score(Y, kmeans.labels_)
    print(k, s_score)

2 0.11425266909422802
3 0.10212519930324153
4 0.10386199279728914
5 0.10262978788826137
6 0.10289606693122941
7 0.10566732976017826
8 0.09318304584849725


### ANSWER: As the scores shown above, when k=2 we could get much higher score. ###

#### 7.b ####
**Question**: Using the the sum of within cluster variance metric with the elbow method what was the best k?  
**Answer**:  

In [88]:
Y = df_5

for k in range(2, 9):
    kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)
    df_5['cluster'] = kmeans.labels_
    df_var = df_5.groupby('cluster').var()
    v_score = df_var.sum(axis=1).sum(axis=0)
    print(k, v_score)

2 65.05446747881629
3 89.15503658809561
4 129.3437461615741
5 123.592215249123
6 197.18459319750903
7 211.63258429552334
8 209.8501379784086


### ANSWER: As the sum of variance shown above, when k=2 we could get the smalest sum of variance, which is the most desired score. ###

### Question 8 ###
For this question please come up with your own question about this dataset and using a clustering technique as part of your method of answering it. Describe in short the question, and how clustering can answer that question.




```
Do dummy variables need to be normalized? Compare the clustering results by within-cluster variance metric
```



In [89]:
df_4 = df_2.drop(columns = ['user_id','q16s','q16t'])
df_4['log15'] = df_4['q15'].apply(lambda x: math.log(x))
df_4 = df_4.drop(columns = ['q15'])

df_6 = pd.concat([df_4, df_3], axis=1)
df_6=(df_6 - df_6.mean()) / df_6.std()

Y = df_6

for k in range(2, 9):
    kmeans = KMeans(n_clusters=k, random_state=0).fit(Y)
    df_6['cluster'] = kmeans.labels_
    df_var = df_6.groupby('cluster').var()
    v_score = df_var.sum(axis=1).sum(axis=0)
    print(k, v_score)

2 78.26918312332558
3 108.22447802596838
4 133.6027555669955
5 165.4258756132174
6 203.79806608386662
7 227.45042819774184
8 351.96632071054546


### ANSWER:  No need. When dummy variables normalized, the metric is worse than not. Therefor, in this test the dummy variables do not need to be normalized. ###

## Bonus question (+20%) - Reviewer overlap:
Create a dataset with f reviewers as the rows and business_ids as the columns (there are a lot) where the feature value is is ‘1’ if the reviewer has written a review for that business and ‘0’ if not. Use the methods described in this assignment to answer the question of how much overlap of businesses reviewed exists among reviewers in this dataset. Print head() of this newly created dataset.

In [1]:
import pandas as pd
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
import numpy as np
from sklearn.preprocessing import normalize

!wget http://people.ischool.berkeley.edu/~zp/course_datasets/yelp_reviews.csv
df_review = pd.read_csv('yelp_reviews.csv')

--2019-02-14 21:35:06--  http://people.ischool.berkeley.edu/~zp/course_datasets/yelp_reviews.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 376638166 (359M) [text/csv]
Saving to: ‘yelp_reviews.csv.5’

yelp_reviews.csv.5  100%[===================>] 359.19M  12.3MB/s    in 43s     

2019-02-14 21:35:49 (8.30 MB/s) - ‘yelp_reviews.csv.5’ saved [376638166/376638166]



In [0]:
df_A = df_review.sample(n = int( len(df_review)*0.3) ) 

df_B = df_A.groupby(['user_id', 'business_id']).size()
df_C = df_B.reset_index().iloc[:,0:2]
df_C['cnt'] = 1
df_D = pd.pivot_table(df_C, values='cnt', index='user_id', columns=['business_id'], aggfunc=np.sum )

In [5]:
df_E = df_D
df_E.fillna(value=0, inplace=True)

print(df_E[df_E['--5jkZ3-nUPZxUvtcbr8Uw']==1])

business_id             --5jkZ3-nUPZxUvtcbr8Uw  --jFTZmywe7StuZ2hEjxyA  \
user_id                                                                  
5kH4RrEyebBr75ASdxnzSg                     1.0                     0.0   
QbMJhsqG1zWlzoOBPFwDnQ                     1.0                     0.0   
tCa2ajxRYbX_RiUX0VPlfA                     1.0                     0.0   

business_id             -0Ackw6MF82PXO9f9Jh_Kg  -0Oh0BEtQEC9OmmzZ_H5Bg  \
user_id                                                                  
5kH4RrEyebBr75ASdxnzSg                     0.0                     0.0   
QbMJhsqG1zWlzoOBPFwDnQ                     0.0                     0.0   
tCa2ajxRYbX_RiUX0VPlfA                     0.0                     0.0   

business_id             -1ERbsOf9XOC9wGbZYFr7g  -1tQu_yKCOgOj-WtXmMYIA  \
user_id                                                                  
5kH4RrEyebBr75ASdxnzSg                     0.0                     0.0   
QbMJhsqG1zWlzoOBPFwDnQ              

In [11]:
sum_business = pd.DataFrame(df_E.sum(axis=0))
sum_business.columns = ['sum']

def fun (x):
    if x['sum'] >= 2 :
      return 1
    elif x['sum'] == 1 :
      return 0
    
sum_business['overlap'] = sum_business.apply(lambda x: fun(x), axis=1)

print(sum_business.head())

                         sum  overlap
business_id                          
--5jkZ3-nUPZxUvtcbr8Uw   3.0        1
--jFTZmywe7StuZ2hEjxyA   1.0        0
-0Ackw6MF82PXO9f9Jh_Kg  15.0        1
-0Oh0BEtQEC9OmmzZ_H5Bg   3.0        1
-1ERbsOf9XOC9wGbZYFr7g   8.0        1
4.4220964896270605


In [12]:

print(sum_business['overlap'].mean())

0.4805621463770761


### ANSWER: As the mean above, the overlap ratio of business is 48.06%. ###

## Bonus question #2 (+10%) - Do you like Boba?

    A. YEAH
    B. NO
    B. HELL YEAH
 
 ### **Justify your answer. Put extra effort into your justification if you choose 'B'.**
    


### ANSWER: For sure A. I am from Taiwan, and Boba is the pride of Taiwanese. I wish everyone enjoys boba very much!!! ###